File: code/nodejs/07-harness/fallback_chain.ts
/**
 * Multi-level LLM provider fallback chain with circuit breakers.
 *
 * Principle: every LLM call needs a fallback path. Providers are tried
 * in priority order; failures trigger automatic promotion to the next level.
 *
 * Circuit breaker states:
 *   CLOSED → OPEN (after threshold failures in window)
 *   OPEN → HALF_OPEN (after cooldown expires)
 *   HALF_OPEN → CLOSED (on success) | OPEN (on failure)
 *
 * See: docs/07-harness-engineering/01-the-harness-mindset.md
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface LLMResponse {
  content: string;
  toolCalls?: unknown[];
  tokensUsed: number;
  model: string;
  providerName: string;
  fallbackLevel: number;
}

export type ProviderFn = (
  messages: Array<{ role: string; content: string }>,
  options: { model: string; maxTokens: number; temperature: number }
) => Promise<LLMResponse>;

export interface FallbackProvider {
  name: string;
  fn: ProviderFn;
  model: string;
  maxTokens: number;
  temperature: number;
  priority: number;     // Lower number = tried first
}

export class AllProvidersFailedError extends Error {
  constructor(public errors: string[]) {
    super(`All providers failed:\n${errors.map((e) => `  - ${e}`).join("\n")}`);
    this.name = "AllProvidersFailedError";
  }
}

// ---------------------------------------------------------------------------
// Circuit breaker
// ---------------------------------------------------------------------------

type CircuitState = "closed" | "open" | "half_open";

export class CircuitBreaker {
  private state: CircuitState = "closed";
  private failureTimestamps: number[] = [];
  private openedAt?: number;

  constructor(
    private threshold = 5,
    private windowMs = 60_000,
    private cooldownMs = 120_000
  ) {}

  get currentState(): CircuitState {
    if (this.state === "open") {
      const elapsed = Date.now() - (this.openedAt ?? 0);
      if (elapsed >= this.cooldownMs) {
        this.state = "half_open";
      }
    }
    return this.state;
  }

  allowRequest(): boolean {
    return this.currentState !== "open";
  }

  recordSuccess(): void {
    this.state = "closed";
    this.failureTimestamps = [];
  }

  recordFailure(): void {
    const now = Date.now();
    this.failureTimestamps = this.failureTimestamps.filter((t) => now - t < this.windowMs);
    this.failureTimestamps.push(now);

    if (this.state === "half_open" || this.failureTimestamps.length >= this.threshold) {
      this.state = "open";
      this.openedAt = now;
    }
  }
}

// ---------------------------------------------------------------------------
// FallbackChain
// ---------------------------------------------------------------------------

export interface FallbackAttempt {
  provider: string;
  fallbackLevel: number;
  success: boolean;
  error?: string;
  durationMs: number;
}

export interface FallbackResult {
  response: LLMResponse;
  attempts: FallbackAttempt[];
  usedFallback: boolean;
  finalLevel: number;
}

export class FallbackChain {
  private circuits = new Map<string, CircuitBreaker>();

  constructor(
    private providers: FallbackProvider[],
    private circuitOptions: { threshold?: number; windowMs?: number; cooldownMs?: number } = {}
  ) {
    const sorted = [...providers].sort((a, b) => a.priority - b.priority);
    this.providers = sorted;
    for (const p of sorted) {
      this.circuits.set(p.name, new CircuitBreaker(
        circuitOptions.threshold,
        circuitOptions.windowMs,
        circuitOptions.cooldownMs
      ));
    }
  }

  async call(
    messages: Array<{ role: string; content: string }>
  ): Promise<FallbackResult> {
    const attempts: FallbackAttempt[] = [];
    const errors: string[] = [];

    for (let level = 0; level < this.providers.length; level++) {
      const provider = this.providers[level];
      const circuit = this.circuits.get(provider.name)!;

      if (!circuit.allowRequest()) {
        attempts.push({
          provider: provider.name,
          fallbackLevel: level,
          success: false,
          error: "Circuit open — skipped",
          durationMs: 0,
        });
        continue;
      }

      const start = Date.now();
      try {
        const response = await provider.fn(messages, {
          model: provider.model,
          maxTokens: provider.maxTokens,
          temperature: provider.temperature,
        });
        circuit.recordSuccess();
        attempts.push({
          provider: provider.name,
          fallbackLevel: level,
          success: true,
          durationMs: Date.now() - start,
        });
        return {
          response: { ...response, fallbackLevel: level },
          attempts,
          usedFallback: level > 0,
          finalLevel: level,
        };
      } catch (err) {
        const errorMsg = err instanceof Error ? err.message : String(err);
        circuit.recordFailure();
        errors.push(`${provider.name}: ${errorMsg}`);
        attempts.push({
          provider: provider.name,
          fallbackLevel: level,
          success: false,
          error: errorMsg,
          durationMs: Date.now() - start,
        });
      }
    }

    throw new AllProvidersFailedError(errors);
  }

  getCircuitStates(): Record<string, string> {
    const states: Record<string, string> = {};
    this.circuits.forEach((cb, name) => {
      states[name] = cb.currentState;
    });
    return states;
  }
}

File: code/go/07-harness/fallback_chain.go
// fallback_chain.go — Cascading fallback chain for LLM provider failures (Go port).
//
// See: docs/07-harness-engineering/02-input-guardrails-and-validation.md
package main

import (
	"context"
	"fmt"
	"time"
)

// ---------------------------------------------------------------------------
// Fallback types
// ---------------------------------------------------------------------------

// FallbackProvider is a single provider in the fallback chain.
type FallbackProvider struct {
	Name       string
	Model      string
	Priority   int // lower = higher priority
	MaxRetries int
	TimeoutMs  int
}

// FallbackResult holds the outcome of a fallback chain attempt.
type FallbackResult struct {
	ProviderUsed string
	Answer       string
	Attempts     int
	TotalElapsed time.Duration
	Error        error
}

// FallbackChain tries providers in priority order until one succeeds.
type FallbackChain struct {
	providers []FallbackProvider
	call      func(ctx context.Context, provider FallbackProvider, prompt string) (string, error)
}

// NewFallbackChain creates a fallback chain with a caller function.
func NewFallbackChain(
	providers []FallbackProvider,
	call func(ctx context.Context, provider FallbackProvider, prompt string) (string, error),
) *FallbackChain {
	// Sort by priority (simple insertion sort)
	sorted := make([]FallbackProvider, len(providers))
	copy(sorted, providers)
	for i := range sorted {
		for j := i + 1; j < len(sorted); j++ {
			if sorted[j].Priority < sorted[i].Priority {
				sorted[i], sorted[j] = sorted[j], sorted[i]
			}
		}
	}
	return &FallbackChain{providers: sorted, call: call}
}

// Execute tries each provider in order, retrying up to MaxRetries times.
func (c *FallbackChain) Execute(ctx context.Context, prompt string) FallbackResult {
	start := time.Now()
	attempts := 0

	for _, provider := range c.providers {
		for retry := 0; retry <= provider.MaxRetries; retry++ {
			attempts++
			answer, err := c.call(ctx, provider, prompt)
			if err == nil {
				return FallbackResult{
					ProviderUsed: provider.Name,
					Answer:       answer,
					Attempts:     attempts,
					TotalElapsed: time.Since(start),
				}
			}
			if retry < provider.MaxRetries {
				fmt.Printf("[Fallback] %s attempt %d failed: %v — retrying\n", provider.Name, retry+1, err)
			}
		}
		fmt.Printf("[Fallback] %s exhausted, trying next provider\n", provider.Name)
	}

	return FallbackResult{
		Attempts:     attempts,
		TotalElapsed: time.Since(start),
		Error:        fmt.Errorf("all providers exhausted after %d attempts", attempts),
	}
}

// DefaultFallbackProviders returns a sensible default provider chain.
func DefaultFallbackProviders() []FallbackProvider {
	return []FallbackProvider{
		{Name: "gpt-4o", Model: "gpt-4o", Priority: 1, MaxRetries: 2, TimeoutMs: 30000},
		{Name: "gpt-4o-mini", Model: "gpt-4o-mini", Priority: 2, MaxRetries: 3, TimeoutMs: 15000},
		{Name: "gpt-3.5-turbo", Model: "gpt-3.5-turbo", Priority: 3, MaxRetries: 3, TimeoutMs: 10000},
	}
}

// RunFallbackChainDemo demonstrates the fallback chain.
func RunFallbackChainDemo() {
	callCount := 0
	chain := NewFallbackChain(DefaultFallbackProviders(), func(ctx context.Context, p FallbackProvider, prompt string) (string, error) {
		callCount++
		if p.Priority <= 2 && callCount <= 4 {
			return "", fmt.Errorf("simulated failure for %s", p.Name)
		}
		return fmt.Sprintf("[%s] Answer to: %s", p.Name, prompt), nil
	})

	result := chain.Execute(context.Background(), "What is 2+2?")
	if result.Error != nil {
		fmt.Printf("Fallback chain failed: %v\n", result.Error)
	} else {
		fmt.Printf("Provider: %s, Attempts: %d, Elapsed: %v\n",
			result.ProviderUsed, result.Attempts, result.TotalElapsed)
		fmt.Printf("Answer: %s\n", result.Answer)
	}
}
